crates/aranya-daemon/src/sync/task/quic.rs

```diff
@@ -150,11 +184,113 @@ impl SyncState for State {
+            .await
+    }
+
+    /// Unsubscribe from hello notifications from a peer.
+    #[instrument(skip_all)]
+    async fn hello_unsubscribe_impl(
+        syncer: &mut Syncer<Self>,
+        id: GraphId,
+        peer: &Addr,
+    ) -> SyncResult<()> {
+        syncer.state.store.set_team(id.into_id().into());
+        syncer
+            .send_hello_unsubscribe_request(peer, id, syncer.server_addr)
+            .await
+    }
+
+    /// Broadcast hello notifications to all subscribers of a graph.
+    #[instrument(skip_all)]
+    async fn broadcast_hello_notifications(
+        syncer: &mut Syncer<Self>,
+        graph_id: GraphId,
+        head: Address,
+    ) -> SyncResult<()> {
+        // Get all subscribers for this graph
+        let subscribers = {
+            let subscriptions = syncer.state.hello_subscriptions.lock().await;
+
+            let filtered: Vec<_> = subscriptions
+                .iter()
+                .filter(|((sub_graph_id, _), _)| *sub_graph_id == graph_id)
+                .map(|((_, addr), subscription)| (*addr, subscription.clone()))
+                .collect();
+
+            filtered
+        };
+
+        // Send hello notification to each subscriber
+        for (subscriber_addr, subscription) in subscribers.iter() {
+            // Check if enough time has passed since last notification
+            if let Some(last_notified) = subscription.last_notified {
+                let delay = Duration::from_millis(subscription.delay_milliseconds);
+                let elapsed = last_notified.elapsed();
+                if elapsed < delay {
+                    continue;
+                }
+            }
+
+            // Send the notification
+            // Convert SocketAddr to Addr for the syncer method
+            let peer_addr = Addr::from(*subscriber_addr);
+            match syncer
+                .send_hello_notification_to_subscriber(&peer_addr, graph_id, head)
+                .await
+            {
+                Ok(()) => {
+                    // Update the last notified time
+                    let mut subscriptions = syncer.state.hello_subscriptions.lock().await;
+                    if let Some(sub) = subscriptions.get_mut(&(graph_id, *subscriber_addr)) {
+                        sub.last_notified = Some(Instant::now());
+                    } else {
+                        warn!(
+                            ?subscriber_addr,
+                            ?graph_id,
+                            "Failed to find subscription to update last_notified"
+                        );
+                    }
+                }
+                Err(e) => {
+                    warn!(
+                        error = %e,
+                        ?subscriber_addr,
+                        ?head,
+                        "Failed to send hello notification"
+                    );
+                }
+            }
```

Comment:
Or use `if let` to simplify:
```suggestion
            if let Err(e) =  syncer
                .send_hello_notification_to_subscriber(&peer_addr, graph_id, head)
                .await
            {
                    warn!(
                        error = %e,
                        ?subscriber_addr,
                        ?head,
                        "Failed to send hello notification"
                    );
                    continue;
            }
            
            // Update the last notified time
            let mut subscriptions = syncer.state.hello_subscriptions.lock().await;
            if let Some(sub) = subscriptions.get_mut(&(graph_id, *subscriber_addr)) {
                sub.last_notified = Some(Instant::now());
            } else {
                warn!(
                    ?subscriber_addr,
                    ?graph_id,
                    "Failed to find subscription to update last_notified"
                );
            }
```